.kitchen.yml

```diff
@@ -23,88 +24,59 @@ driver_config:
 provisioner:
   name: chef_zero
   product_name: chef
-  product_version: 14
+  product_version: 15
   install_strategy: once
   data_path: test
 
 platforms:
-- name: centos-7.6
+- name: centos-8.3
   attributes:
+    elasticsearch:
+      download_url: "https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-%s-x86_64.rpm"
+      download_checksum: "aa10954329d659d7d3a2f8f1abfe3be764b09c44a6248b7a25d6b95c38c03fe3"
     mongodb:
       reload_action: "restart"
-- name: ubuntu-16.04
+- name: ubuntu-20.04
+  attributes:
+    elasticsearch:
+      download_url: "https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-%s-amd64.deb"
+      download_checksum: "b91564a4ebf926c7634b5817e7b8a9f82335a0554a2a15740b3a8f42d2bfff8f"
   run_list:
   - recipe[apt]
-- name: debian-9
+- name: debian-10.10
+  attributes:
+    elasticsearch:
+      download_url: "https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-%s-amd64.deb"
+      download_checksum: "b91564a4ebf926c7634b5817e7b8a9f82335a0554a2a15740b3a8f42d2bfff8f"
   run_list:
   - recipe[apt]
 
 suites:
-  - name: oracle
-    attributes:
-      machine_fqdn: graylog.local
-      machine_fqdn_as_hostname: true
-      apt:
-        compile_time_update: true
-      java:
-        jdk_version: '8'
-        install_flavor: 'oracle'
-        oracle:
-          accept_oracle_download_terms: true
-      mongodb:
-        package_version: '4.0.6'
-      graylog2:
-        password_secret: "2jueVqZpwLLjaWxV"
-        # Credentials are admin:admin
-        root_password_sha2: "8c6976e5b5410415bde908bd4dee15dfb167a9c873fc4bb8a81f6f2ab448a918"
-        restart: "delayed"
-        sidecar:
-          release: '0.1.8'
-          version: '0.1.8'
-          build: 1
-      elasticsearch:
-        version: "5.6.14"
-        cluster:
-          name: "graylog"
-    run_list:
-      - recipe[fqdn]
-      - recipe[curl]
-      - recipe[sc-mongodb]
-      - recipe[java]
-      - recipe[elasticsearch_test]
-      - recipe[graylog2]
-      - recipe[graylog2::server]
-      - recipe[graylog2::collector_sidecar]
-      - recipe[graylog2::authbind]
   - name: openjdk
     attributes:
       machine_fqdn: graylog.local
       machine_fqdn_as_hostname: true
       apt:
         compile_time_update: true
-      java:
-        jdk_version: '8'
-        install_flavor: 'openjdk'
```

Comment:
I think we need to set these values, otherwise a test run fails with:

```
       Recipe Compile Error in /tmp/kitchen/cache/cookbooks/graylog2/recipes/default.rb
         ================================================================================
         
         NoMethodError
         -------------
         undefined method `[]' for nil:NilClass
         
         Cookbook Trace:
         ---------------
           /tmp/kitchen/cache/cookbooks/graylog2/recipes/default.rb:2:in `from_file'
         
         Relevant File Content:
         ----------------------
         /tmp/kitchen/cache/cookbooks/graylog2/recipes/default.rb:
         
           1:  version = node['graylog2']['major_version']
           2>> raise('Java version needs to be >= 8') if node['java']['jdk_version'].to_i < 8
```

Reply:
The only problem is, running `recipe[java]` doesn't work anymore. Running the default recipe results in an error that says you have to define it as a custom resource. So those attributes don't actually have an effect on what version of Java is actually installed. 

Is it still a concern that someone might be running older versions of Java? I removed the test for now, but let me know. Maybe there is some other way to check the version?

Reply:
Graylog is quite picky about the used Java version. If I am right we support 8-10, but currently Java is at 18. This is not obvious for the user, so checking the version in one way or another would be nice.

Reply:
It runs okay on java 11. I have a PR to make a Java 11 version of the docker image. And I remember we updated the graylogctl script so that it would run on Java 15. 

I'll see if I can figure out how the best way to check the Java version.

Reply:
It looks like ohai actually collects this information for us. So I should be able to use `#{node['languages']['java']['version']}`.

Reply:
Alright, I added a different Java test here. What do you think?

Reply:
Looks neat! Upper version check as well? Usually, folks tend to install the latest Java. `8 < && > 11`

Reply:
Okay, I changed it around a bit.

Reply:
Check looks good now. It's just a chicken/egg problem now. Java is not installed anymore therefore no Ohai fact exists. :/

Reply:
Yeah, I'm struggling with the rspec tests. It keeps failing with:

`undefined method `[]' for nil:NilClass1`.

I'm guessing `node' isn't populated for rspec tests? Not sure how it works.

Reply:
Yes, get the same. `node['languages']` is defined, but because no Java is installed, `node['languages']['java']` is nil.
At least for the tests we need to install Java somehow. The latest Java cookbook is looking a bit different:
https://github.com/sous-chefs/java/blob/main/documentation/resources/openjdk_install.md#examples

Reply:
With the last commit, this seams to be fixed :+1:

Reply:
I think I have a way of making sure the tests install Java. I'm still experimenting with it.

Reply:
Yeah, I'm still playing with it. But it's going in the right direction.

Reply:
The `kitchen setup` isn't working for me anymore using `#{node['languages']['java']['version']}`.

I think java must have a compile phase and a deploy phase. In the .kitchen.yml, I have my java recipe specified. But it never gets run. It always goes straight to compiling the graylog default.rb, and then fails because java isn't installed, and so the metadata isn't populated. 

If java is installed to the node from a previous run, then it does work. But if you create a fresh node, it doesn't work.

If I put something like `java_version = shell_out("java -version").stdout` into default.rb, it still skips the java recipe install and goes straight into graylog's default.rb and complains that it can't find java. 

The only thing that works is specifying the custom java resource in default.rb, like this:

`openjdk_install "#{node['java']['jdk_version']}"`

That means we are installing openjdk ourselves. But the problem there is, the user might not want openjdk. They might want oracle jdk or zulu jdk. Or something else.

Reply:
I would prefer that users have the freedom to install Java themselves. They may be installing this recipe onto a server that already has Java on it.

Reply:
They way I had this working before is that I created a new recipe, then I put `openjdk_install '11`` into it, then I only specified that recipe in .kitchen.yml.

Reply:
It makes me wonder, if we want java to be a dependency, we may have to install it ourselves. I just don't like that. Maybe I can put options into my java recipe for different flavors of java. But I only ever intended the recipe to be used for tests.

Reply:
Yes, totally agree. For normal operation, there should be no hard dependency on a specific Java flavor. But the tests should be self-contained imo. We can look for a Vagrant box with pre-installed Java but I can also take another look into the tests. Maybe there is a nice way of getting a wrapper cookbook in.

Reply:
I have pushed a commit with a wrapper cookbook for the integration tests. In that way, we don't need graylog2::java. That makes it more clear that the user has to install Java on its own. You can take a look and also revert if you don't like it :)

Reply:
That looks great! This shows my inexperience with Chef. I see in the `ruby_block` documentation, it says:

```
Ruby code in the ruby_block resource is evaluated with other resources during convergence, whereas Ruby code outside of a ruby_block resource is evaluated before other resources, as the recipe is compiled.```

So that must be what the problem was. I didn't realize it worked that way.

Reply:
Okay, that worked. I guess I need to fix the rspec tests now!